File: Lite/src/python/lite/Template.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from Expression import evaluate
from StringIO import StringIO

EL_TYPE = 0;            #// [0,<el>]
IF_TYPE = 1;            #// [1,[...],<test el>]
BREAK_TYPE = 2;         #// [2,depth]
XML_ATTRIBUTE_TYPE = 3; #// [3,<value el>,'name']
XML_TEXT_TYPE = 4;      #// [4,<el>]
FOR_TYPE = 5;           #// [5,[...],<items el>,'varName']
ELSE_TYPE = 6;          #// [6,[...],<test el>] //<test el> 可为null
ADD_ON_TYPE = 7;        #// [7,[...],<add on el>,'<addon-class>']
VAR_TYPE = 8;           #// [8,<value el>,'name']
CAPTRUE_TYPE = 9;       #// [9,[...],'var']

FOR_KEY = "for"
IF_KEY = "if"

class Template:
    items = None
    def __init__(self, list):
        self.items = list
        
    def render(self, context, out):
        renderList(context, self.items, out)
def renderList(context, children, out):
    for item in children:
        try:
            if isinstance(item, str):
                out.write(item)
            else:
                if item[0] == EL_TYPE:
                    processExpression(context, item, out, False)
                elif item[0] == XML_TEXT_TYPE:
                    processExpression(context, item, out, True)
                elif item[0] == VAR_TYPE:
                    processVar(context, item)
                elif item[0] == CAPTRUE_TYPE:
                    processCaptrue(context, item)
                elif item[0] == IF_TYPE:
                    processIf(context, item, out)
                elif item[0] == ELSE_TYPE:
                    processElse(context, item, out)
                elif item[0] == FOR_TYPE:
                    processFor(context, item, out)
                elif item[0] == XML_ATTRIBUTE_TYPE:
                    processAttribute(context, item, out)
        except (Exception, ), e:
            #print item[0]
            out.write(str(e))
            pass

def printXMLAttribute(text, out):
    if isinstance(text,bool):
        out.write(text and 'true' or 'false');
    else:
	    for c in str(text):
	        if c == '<':
	            out.write("&lt;")
	            break
	        elif c == '>':
	            out.write("&gt;")
	            break
	        elif c == '&':
	            out.write("&amp;")
	            break
	        elif c == '"':
	            out.write("&#34;")
	            break
	        else:
	            out.write(c)

def printXMLText(text, out):
    if isinstance(text,bool):
        out.write(text and 'true' or 'false');
    else:
	    for c in str(text):
	        if c == '<':
	            out.write("&lt;")
	            break
	        elif c == '>':
	            out.write("&gt;")
	            break
	        elif c == '&':
	            out.write("&amp;")
	            break
	        else:
	            out.write(c)

def toBoolean(test):
    if test is None:
        return False
    else:
        if isinstance(test, list) or isinstance(test, dict):
            return true
        elif test:
            return True
        else:
            return False
    return True

def processExpression(context, data, out, encodeXML):
    value = evaluate(data[1],context)
    if encodeXML and value is not None:
        printXMLText(value, out)
    else:
        out.write(value)

def processIf(context, data, out):
    test = True
    try:
        if toBoolean(evaluate(data[2],context)):
            renderList(context, data[1], out)
        else:
            test = True
    finally:
        context[IF_KEY]=test

def processElse(context, data, out):
    if not toBoolean(context[IF_KEY]):
        try:
            if data[2] is None or toBoolean(evaluate(data[2],context)):
                renderList(context, data[1], out)
                context[IF_KEY] = True
        except (Exception, ), e:
            #if log.isDebugEnabled():
            #    log.debug(e)
            context[IF_KEY] = True

def processFor(context, data, out):
    children = data[1]
    items = evaluate(data[2],context)
    varName = data[3]
    length = len(items)
    preiousStatus = hasattr(context,FOR_KEY) and context[FOR_KEY]
    try:
        forStatus = ForStatus(length)
        context[FOR_KEY]=forStatus
        for item in items:
            forStatus.index += 1
            context[varName]=item
            renderList(context, children, out)
    finally:
        context[FOR_KEY]=preiousStatus
        context[IF_KEY]= length > 0

def processVar(context, data):
    context[data[2]]= evaluate(data[1],context);

def processCaptrue(context, data):
    buf = StringIO();
    renderList(context, data[1], buf)
    context[data[2]]= buf.getvalue();

def processAttribute(context, data, out):
    result = evaluate(data[1],context)
    if data[2] is None:
        printXMLAttribute(result, out)
    elif result is not None:
        out.write(" ")
        out.write(data[2])
        out.write("=\"")
        printXMLAttribute(result, out)
        out.write('"')

class ForStatus(object):
    """ generated source for ForStatus

    """
    index = -1
    lastIndex = 0

    def __init__(self, end):
        self.lastIndex = end - 1

File: Lite/src/php/lite/Template.php
<?php
require_once('Expression.php');
define('LITE_FOR_KEY', 'for');
define('LITE_IF_KEY', 'if');

class Template{

    function Template($items){
        $this->items = &$items;
    }
    function render($context){
        lite_render_list($context, $this->items);
    }
}
function lite_render_list(&$context, &$children){
    foreach($children as &$item){
        try{
            if(is_string($item)){
                echo $item;
            }else{
                switch($item[0]){
                	case 0://LITE_EL_TYPE:
                    	lite_process_el($context, $item, 0);
                    	break;
	                case 1://LITE_IF_TYPE:
	                    lite_process_if($context, $item);
                    	break;
	                case 3://LITE_XML_ATTRIBUTE_TYPE:
	                    lite_process_attribute($context, $item);
                    	break;
	                case 4://LITE_XML_TEXT_TYPE:
	                    lite_process_el($context, $item, 1);
                    	break;
	                case 5://LITE_FOR_TYPE:
	                    lite_process_for($context, $item);
                    	break;
	                case 6://LITE_ELSE_TYPE:
	                    lite_process_else($context, $item);
                    	break;
	                case 8://LITE_VAR_TYPE:
	                    lite_process_var($context, $item);
                    	break;
	                case 9://LITE_CAPTRUE_TYPE:
	                    lite_process_captrue($context, $item);
                    	break;
	            }
	        }
        }catch(Exception $e){
            echo $e;
        }
    }
}
function lite_print_xml_attribute(&$text){
    lite_print_xml_text($text);
}

function lite_print_xml_text(&$text){
    if(is_bool($text)){
        echo $text?'true':'false';
    }else{
        echo htmlspecialchars($text);
    }
}

function lite_to_bool(&$test){
    return !!$test;
}

function lite_process_el(&$context, &$data, $encodeXML){
	$stack = &$data[1];
    $value = lite_evaluate($context,$stack);
    if($encodeXML && isset($value)){
        lite_print_xml_text($value);
    }else{
        echo $value;
    }
}
function lite_process_if(&$context, &$data){
    try{
        if(lite_to_bool(lite_evaluate($context,$data[2]))){
        	$test = 1;
            lite_render_list($context, $data[1]);
        }else{
            $test = 0;
        }
        $context[LITE_IF_KEY]=$test;
    }catch(Exception $e){
        $context[LITE_IF_KEY]=$test;
        throw $e;
    }
}
function lite_process_else(&$context, &$data){
    if(array_key_exists(LITE_IF_KEY,$context) && !lite_to_bool($context[LITE_IF_KEY])){
        try{
            if(is_null($data[2]) || lite_to_bool(lite_evaluate($context,$data[2]))){
                $test = 1;
                lite_render_list($context, $data[1]);
            }else{
            	$test = 0;
            }
            $context[LITE_IF_KEY] = $test;
        }catch (Exception $e){
            $context[LITE_IF_KEY] = $test;
            throw $e;
        }
    }
}
function lite_process_for(&$context, &$data){
    $children = &$data[1];
    $items = lite_evaluate($context,$data[2]);
    $varName = &$data[3];
    if(array_key_exists(LITE_FOR_KEY,$context)){
    	$preiousStatus =  &$context[LITE_FOR_KEY];
    }
    try{
        $length = 0;
        $forStatus = array('index'=>-1);
        $context[LITE_FOR_KEY]=&$forStatus;
        if(is_numeric($items)){
        	$i=0;
        	$length = $items;
        	$forStatus['lastIndex']=$length-1;
	        while($i<$length){
	            $forStatus['index'] = $i++;
	            $context[$varName] = $i;
	            lite_render_list($context, $data);
	        }
        }else if(is_array($items)){
            $length = count($items);
        	$forStatus['lastIndex'] = $length-1;
	        foreach($items as &$item){
	            $forStatus['index'] += 1;
	            $context[$varName] = $item;
	            lite_render_list($context, $children);
	        }
        }
        $context[LITE_FOR_KEY]=&$preiousStatus;
        $context[LITE_IF_KEY]= $length > 0;
    }catch(Exception $e){
        $context[LITE_FOR_KEY]=$preiousStatus;
        $context[LITE_IF_KEY]= $length > 0;
        throw $e;
    }
}

function lite_process_var(&$context, &$data){
    $context[$data[2]]= lite_evaluate($context,$data[1]);
} 

function lite_process_captrue(&$context, &$data){
    ob_start();
    lite_render_list($context, $data[1]);
    $context[$data[2]]= ob_get_contents();
    ob_end_clean();
}

function lite_process_attribute(&$context, &$data){
    $result = lite_evaluate($context,$data[1]);
    if(is_null($data[2])){
        lite_print_xml_attribute($result);
    }elseif(isset($result)){
        echo ' ';
        echo $data[2];
        echo '="';
        lite_print_xml_attribute($result);
        echo '"';
    }
}
?>